### PR TITLE
Made the APIs of the Tensorboard Logger and WandB Logger same

### LINKED ISSUE #2791: WandBLogger and TensorboardLogger have different APIs for logging audio
## 🚀 Feature

The following code doesn't work:

```
logger = WandBLogger()
logger.writer
```

This is how you would typically add audio with a tensorboard logger:
```
logger.writer.add_audio('mixture', x.t(), engine.state.epoch)
```

The workaround (similar to discussed in https://github.com/Lightning-AI/lightning/issues/7028) would be to use the underlying _wandb object:
```
logger._wandb.log({"validation": [wandb.Audio(x.t(), caption="mixture", sample_rate=44100)]})
logger._wandb.log({"validation": [wandb.Audio(x.t(), caption="vocals", sample_rate=44100)]})
```

Is there a proposal for a standardized media logging API?

### PR DESCRIPTION
**Pull request by guptaaryan16**
Fixes #2791 

Description: Added a __getattr__ method to the Tensorboard file

Check list:

- [X] New tests are added (if a new feature is added)
- [ ] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

File: ignite/contrib/handlers/tensorboard_logger.py [hunk 1]
```diff
@@ -160,8 +160,11 @@ def __init__(self, *args: Any, **kwargs: Any):
 
         self.writer = SummaryWriter(*args, **kwargs)
 
+    def __getattr__(self, attr: Any) -> Any:
+        return getattr(self.writer, attr)
+
     def close(self) -> None:
-        self.writer.close()
+        self.close()
```
Comment (vfdev-5) [inline ignite/contrib/handlers/tensorboard_logger.py:167]: I think you should keep previous implementation here

File: tests/ignite/contrib/handlers/test_tensorboard_logger.py [hunk 2]
```diff
@@ -37,20 +37,20 @@ def test_optimizer_params():
     optimizer = torch.optim.SGD([torch.tensor(0.0)], lr=0.01)
     wrapper = OptimizerParamsHandler(optimizer=optimizer, param_name="lr")
     mock_logger = MagicMock(spec=TensorboardLogger)
-    mock_logger.writer = MagicMock()
+    mock_logger = MagicMock()
```
Comment (vfdev-5) [inline tests/ignite/contrib/handlers/test_tensorboard_logger.py:40]: Same here, revert this line please

File: tests/ignite/contrib/handlers/test_tensorboard_logger.py [hunk 3]
```diff
@@ -37,20 +37,20 @@ def test_optimizer_params():
...
     mock_engine.state.iteration = 123
 
     wrapper(mock_engine, mock_logger, Events.ITERATION_STARTED)
-    mock_logger.writer.add_scalar.assert_called_once_with("lr/group_0", 0.01, 123)
+    mock_logger.add_scalar.assert_called_once_with("lr/group_0", 0.01, 123)
 
     wrapper = OptimizerParamsHandler(optimizer, param_name="lr", tag="generator")
     mock_logger = MagicMock(spec=TensorboardLogger)
-    mock_logger.writer = MagicMock()
+    mock_logger = MagicMock()
```
Comment (vfdev-5) [inline tests/ignite/contrib/handlers/test_tensorboard_logger.py:50]: Same here, revert this line please

File: tests/ignite/contrib/handlers/test_tensorboard_logger.py [hunk 4]
```diff
@@ -67,7 +67,7 @@ def test_output_handler_output_transform():
 
     wrapper = OutputHandler("tag", output_transform=lambda x: x)
     mock_logger = MagicMock(spec=TensorboardLogger)
-    mock_logger.writer = MagicMock()
+    mock_logger = MagicMock()
```
Comment (vfdev-5) [inline tests/ignite/contrib/handlers/test_tensorboard_logger.py:70]: Same here, revert this line please and also below

File: tests/ignite/contrib/handlers/test_tensorboard_logger.py [hunk 5]
```diff
@@ -677,7 +677,7 @@ def test_no_torch_utils_tensorboard_package(dirname):
 
     with patch.dict("sys.modules", {"torch.utils.tensorboard": None}):
         tb_logger = TensorboardLogger(log_dir=dirname)
-        assert isinstance(tb_logger.writer, SummaryWriter), type(tb_logger.writer)
+        assert isinstance(tb_logger, SummaryWriter), type(tb_logger)
```
Comment (vfdev-5) [inline tests/ignite/contrib/handlers/test_tensorboard_logger.py:680]: Same here, revert this line please and above

**Review by vfdev-5 — COMMENTED** [review]
@guptaaryan16 thanks for a PR, I left few comments how to fix it

File: tests/ignite/contrib/handlers/test_tensorboard_logger.py [hunk 6]
```diff
@@ -37,20 +37,20 @@ def test_optimizer_params():
...
     mock_logger = MagicMock(spec=TensorboardLogger)
-    mock_logger = MagicMock()
+    mock_logger.writer = MagicMock()
     mock_engine = MagicMock()
     mock_engine.state = State()
     mock_engine.state.iteration = 123
 
     wrapper(mock_engine, mock_logger, Events.ITERATION_STARTED)
-    mock_logger.add_scalar.assert_called_once_with("lr/group_0", 0.01, 123)
+    mock_logger.writer.add_scalar.assert_called_once_with("lr/group_0", 0.01, 123)
```
Comment (vfdev-5) [inline tests/ignite/contrib/handlers/test_tensorboard_logger.py:46]: In the test we still however want to test `__getattr__`, so let's keep `mock_logger.add_scalar` instead of `mock_logger.writer.add_scalar` and others where we call `mock_logger.writer.some_func` -> `mock_logger.some_func`

File: tests/ignite/contrib/handlers/test_tensorboard_logger.py [hunk 7]
```diff
@@ -312,7 +312,7 @@ def test_weights_scalar_handler_whitelist(dummy_model_factory):
 
     wrapper(mock_engine, mock_logger, Events.EPOCH_STARTED)
     mock_logger.writer.add_scalar.assert_called_once_with("weights_norm/fc2/weight", 12.0, 5)
-    mock_logger.writer.reset_mock()
+    mock_logger.reset_mock()
```
Comment (vfdev-5) [inline tests/ignite/contrib/handlers/test_tensorboard_logger.py:315]: Let's revert this line as well

File: tests/ignite/contrib/handlers/test_tensorboard_logger.py (hunk 6 above)
Comment (guptaaryan16) [reply tests/ignite/contrib/handlers/test_tensorboard_logger.py:46]: Ok let me make the change

Comment (guptaaryan16) [reply tests/ignite/contrib/handlers/test_tensorboard_logger.py:46]: Hey I did that and it failed

Comment (vfdev-5) [reply tests/ignite/contrib/handlers/test_tensorboard_logger.py]: OK, mock does not have `__getattr__` replicated from the spec. Let's find another way to test that user can access methods like `writer.add_scalar` directly using `logger.add_scalar`. Do you have any ideas or suggestions ?

Comment (guptaaryan16) [reply tests/ignite/contrib/handlers/test_tensorboard_logger.py]: I guess we can have a multiple inheritance in the tensorboard logger such that it supports all method without the writer instance variable

Comment (guptaaryan16) [reply tests/ignite/contrib/handlers/test_tensorboard_logger.py]: here is the __dict__ of tensorboard logger when I added SummaryWriter as second inheritance in TensorboardLogger class .
{'_mock_return_value': sentinel.DEFAULT, '_mock_parent': None, '_mock_name': None, '_mock_new_name': '', '_mock_new_parent': None, '_mock_sealed': False, '_spec_class': <class 'ignite.contrib.handlers.tensorboard_logger.TensorboardLogger'>, '_spec_set': None, '_spec_signature': <Signature (*args: Any, **kwargs: Any)>, '_mock_methods': ['_SummaryWriter__append_to_scalar_dict', '__abstractmethods__', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__enter__', '__eq__', '__exit__', '__format__', '__ge__', '__getattr__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_abc_impl', '_check_caffe2_blob', '_create_opt_params_handler', '_create_output_handler', '_encode', '_get_comet_logger', '_get_file_writer', 'add_audio', 'add_custom_scalars', 'add_custom_scalars_marginchart', 'add_custom_scalars_multilinechart', 'add_embedding', 'add_figure', 'add_graph', 'add_graph_deprecated', 'add_histogram', 'add_histogram_raw', 'add_hparams', 'add_image', 'add_image_with_boxes', 'add_images', 'add_mesh', 'add_onnx_graph', 'add_openvino_graph', 'add_pr_curve', 'add_pr_curve_raw', 'add_scalar', 'add_scalars', 'add_text', 'add_video', 'attach', 'attach_opt_params_handler', 'attach_output_handler', 'close', 'export_scalars_to_json', 'flush'], '_spec_asyncs': [], '_mock_children': {'writer': <MagicMock name='mock.writer' id='4499319648'>}, '_mock_wraps': None, '_mock_delegate': None, '_mock_called': False, '_mock_call_args': None, '_mock_call_count': 0, '_mock_call_args_list': [], '_mock_mock_calls': [], 'method_calls': [], '_mock_unsafe': False, '_mock_side_effect': None, 'writer': <MagicMock name='mock.writer' id='4499319648'>}

Comment (guptaaryan16) [reply tests/ignite/contrib/handlers/test_tensorboard_logger.py]: I guess through this we can remove the need of writer.add_scalar

Comment (vfdev-5) [reply tests/ignite/contrib/handlers/test_tensorboard_logger.py]: Yes we could inherit also from `SummaryWriter` but this would imply `tensorboard` package as static ignite dependency vs right now it is a runtime dependency and TensorboardLogger class is exposed in ignite even if `tensorboard` package is not installed...

I think we can proceed with `__getattr__` solution and just add a specific test such that we can ensure that methods `logger.add_scalar` is propagated to `logger.writer.add_scalar`.

Comment (guptaaryan16) [reply tests/ignite/contrib/handlers/test_tensorboard_logger.py]: I guess you are right, get_attr is returning the self.writer methods but  Magic block doesn't seem to recognise it so when I printed the __dict__ of a tensorboardlogger I got the dict of the SummaryWriter

**Comment by guptaaryan16** [comment]
I think this will work @vfdev-5

File: tests/ignite/contrib/handlers/test_tensorboard_logger.py [hunk 8]
```diff
@@ -1,5 +1,6 @@
 import math
 import os
+import unittest
```
Comment (vfdev-5) [inline tests/ignite/contrib/handlers/test_tensorboard_logger.py:3]: We are using pytest so let's stick to their formalism without creating TestCase classes and just directly defining test functions.

File: tests/ignite/contrib/handlers/test_tensorboard_logger.py [hunk 9]
```diff
@@ -32,6 +33,15 @@ def test_optimizer_params_handler_wrong_setup():
         handler(mock_engine, mock_logger, Events.ITERATION_STARTED)
 
 
+class TestTensorboardLogger(unittest.TestCase):
+    @patch("ignite.contrib.handlers.TensorboardLogger")
+    def test_add_scalar(self, mock_writer):
+        # Create a TensorboardLogger instance
+        logger = TensorboardLogger()
+        # Call the add_scalar method
+        logger.add_scalar("loss", 0.5)
```
Comment (vfdev-5) [inline tests/ignite/contrib/handlers/test_tensorboard_logger.py:42]: We need to check that `logger.writer.add_scalar` was called with `("loss", 0.5)`

**Review by vfdev-5 — APPROVED** [review]
Thanks @guptaaryan16 , lgtm !
